virtual-desktop/src/app/window-manager/mvd-window-manager/personalization-panel/personalization/personalization.component.ts

```diff
@@ -25,30 +27,283 @@ import { Colors } from '../../shared/colors';
+          textColor = Colors.COOLGREY_90
+        }
+      }
+      this.selectedLightness = lightness;
+      this.desktopThemeService.changeColor(newColor, textColor);
+    }
+  }
+
+  clickSlider($event: any) {
+    console.log($event);
+    this.sliderImgData;
+    let index = ($event.layerY * this.sliderImgData.width + $event.layerX) * 4;
+    let red = this.sliderImgData.data[index];
+    let green = this.sliderImgData.data[index+1];
+    let blue = this.sliderImgData.data[index+2];
+    let hexColor = this.rGBtoHex(red, green, blue);
+
+    // TODO: Adjust lightness of text color based on color theory/contrast/relative luminance
+    let textColor = Colors.COOLGREY_10;
+
+    this.selectedCircle = -1;
+    this.selectedLightness = -1;
+    let hslColor = this.hexToHSL(hexColor);
+    if (hslColor != null) {
+      this.updateLightnessSwatches(hslColor);
+    }
+    this.selectedColor = hexColor;
+    this.desktopThemeService.changeColor(hexColor, textColor);
+  }
+
+  spawnCircles(colors: string[]): void {    
+    var index: number;
+    var canvasName: string;
+    var canvasElem: ElementRef;
+    var numOfCircles = colors.length;
+
+    for (index = 1; index <= numOfCircles; index++) {
+      canvasName = CIRCLE_NAME + index;
+      canvasElem = (<any>this)[canvasName]; // We typecast this to 'any' to avoid silly TS compile problems
+      
+      var context = canvasElem.nativeElement.getContext('2d');
+      var centerX = canvasElem.nativeElement.width / 2;
+      var centerY = canvasElem.nativeElement.height / 2;
     
-    if (color.hsl.l >= .65) { // If lightness of color is too high, we change the text to be dark
-      textColor = Colors.COOLGREY_90
+      var radius = 20;
+
+      if (context != null) {
+        context.beginPath();
+        context.arc(centerX, centerY, radius, 0, 2 * Math.PI, false);
+        context.fillStyle = colors[index-1];
+        context.fill();
+        context.stroke();
+      }
     }
-    this.desktopThemeService.changeColor(color.hex, textColor);
+  }
+
+  spawnSlider(): void {
+    var canvasName: string = SLIDER_NAME + '1';
+    var canvasElem: ElementRef = (<any>this)[canvasName]; // We typecast this to 'any' to avoid silly TS compile problems
+
+    // Create gradient
+    var ctx = canvasElem.nativeElement.getContext("2d");         
+    var gradient = ctx.createLinearGradient(14, 0, 389, 0);
+    gradient.addColorStop(0, 'red');
+    gradient.addColorStop(1 / 6, 'orange');
+    gradient.addColorStop(2 / 6, 'yellow');
+    gradient.addColorStop(3 / 6, 'green');
+    gradient.addColorStop(4 / 6, 'blue');
+    gradient.addColorStop(5 / 6, 'indigo');
+    gradient.addColorStop(1, 'violet');
+
+    // Fill
+    ctx.fillStyle = gradient;
+    ctx.fillRect(0, 0, 389, 14);
+
+    // Save imgData
+    this.sliderImgData = ctx.getImageData(0, 0, canvasElem.nativeElement.width, canvasElem.nativeElement.height);
```

Comment:
We save imgData instead of using getImgData to do a lookup every-time because of performance